File: projets/evaluations/2020_07_02/part3_fruits_implementation.js
/*
Exemple d'implémentation simplifiée en JS (sans contrôle sur les données)
*/



class Fruit
{
    /** @var string label */
    #label;

    /** @var Number price */
    #price;

    getLabel() 
    {
        return this.#label;
    }

    getPrice()
    {
        return this.#price;
    }

    setPrice(price)
    {
        this.#price = price;
    }
}



class Pear extends Fruit
{
    /** @var string color */
    #color;

    getColor()
    {
        return this.#color:
    }

    setColor(color)
    {
        this.#color = color;
    }
}



class Apple extends Fruit 
{
    /** @var bool redLabel */
    #redlabel;

    isRedLabel()
    {
        return this.#redlabel;
    }
}



class Customer
{
    /** @var Number clientId */
    #clientId;

    /** @var Fruit sample */
    #sample;

    constructor(clientId)
    {
        this.#clientId = clientId;
        this.#sample = null;
    }

    getClientId()
    {
        return this.#clientId;
    }
}



class Basket
{
    /** @var Number ref */
    #ref; 

    /** @var Fruit[] items */
    #items;

    constructor(ref, fruit)
    {
        this.#ref = ref;
        this.#items = [];
        this.#items.push(fruit) // Composition = at least 1 Fruit in basket
    }

    getTotalPrice()
    {
        let totalPrice;

        for(let fruit of this.#items) {
            totalPrice += fruit.getPrice();
        }

        return totalPrice;
    }
}

File: projets/evaluations/2020_07_02/part3_fruits_implementation.cs
/*
Exemple d'implémentation simplifiée en C#
*/
class Fruit
{
    private string label;

    private float price;

    public string getLabel()
    {
        return this.label;
    }

    public float getPrice()
    {
        return this.price;
    }

    public void setPrice(float price)
    {
        this.price = price;
    }
}


class Pear : Fruit
{
    private string color;

    public string GetColor() // With C#, all methods use PascalCase
    {
        return this.color;
    }

    public void SetColor(string color)
    {
        this.color = color;
    }
}

class Apple : Fruit 
{
    private bool redlabel;

    public bool IsRedLabel()
    {
        return this.redLabel;
    }
}

class Customer
{
    private int clientId;

    private Fruit sample;

    public Customer(int clientId)
    {
        this.clientId = clientId;
        this.sample = default(Fruit); // default() returns default value for given Type. Our Custom classes are <Nullable>, in this case, default(Fruit) == null.
    }

    public int GetClientId()
    {
        return this.clientId;
    }
}

class Basket
{
    private int @ref; // ref = C# reserved keyword. "@" prefix allows to use a keyword as variable or class attribute name

    private List<Fruit> items;

    public Basket(int _ref, Fruit fruit)
    {
        this.items = new List<Fruit>();
        this.items.Add(fruit); // Composition = at least 1 Fruit in basket
    }

    public float getTotalPrice()
    {
        float totalPrice = 0;

        foreach (Fruit fruit in this.items)
        {
            totalPrice += fruit.GetPrice();
        }

        return totalPrice;
    }

}